### PR TITLE
Shopping cartv2

### PR DESCRIPTION
**Pull request by Eidenn005**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
 	<name>online-book-store</name>
 	<description>Demo project for Spring Boot</description>
 	<properties>
-		<java.version>21</java.version>
+<!--		<java.version>21</java.version>-->
```
Comment (karma-o) [inline pom.xml]: let's uncomment this

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+
+@Entity
+@Getter
+@Setter
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java:20]: do not forget about equals, hashcode & toString, but exclude lazy fields from it, same for all other entity classes

File: src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    @Transactional
+    public CartItemResponseDto updateCartItemQuantity(
+            Long id,
+            CartItemRequestDto requestDto,
+            User user
+    ) {
+        CartItem cartItem = cartItemRepository.findCartItemByBookId(requestDto.getBookId());
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItem = cartItemRepository.save(cartItem);
+        return cartItemMapper.toDto(cartItem);
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java]: this user param is unused, let's use it to validate if this cart item and the cart actually belong to the user who is making the request

File: src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    @Transactional
+    public CartItemResponseDto updateCartItemQuantity(
+            Long id,
+            CartItemRequestDto requestDto,
+            User user
+    ) {
+        CartItem cartItem = cartItemRepository.findCartItemByBookId(requestDto.getBookId());
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java]: this `findCartItemByBookId()` does not make much sense because different users can have the same book in their cart, and you do not have a way differentiate them in case there are several cart items with the same book id, `id` method parameter is also unused, so let's fetch the item directly by it's id, instead of book id

File: src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        return cartItemMapper.toDto(cartItem);
+    }
+
+    @Override
+    @Transactional
+    public void delete(CartItemRequestDto requestDto) {
+        if (cartItemRepository.existsById(requestDto.getBookId())) {
+            cartItemRepository.delete(cartItemRepository
+                    .findCartItemByBookId(requestDto.getBookId()));
+        }
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java]: should we just ignore it if it does not exist? Also let's not put the main flow and logic inside the if body, and stick to the [return early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8)

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private static final Logger logger = LoggerFactory.getLogger(ShoppingCartServiceImpl.class);
+
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto create(User user) {
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    public ShoppingCartDto createForUser(User user) {
```

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            throw new EntityNotFoundException("Not found!");
+        }
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto save(ShoppingCart shoppingCart) {
+        shoppingCart = shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: is it used anywhere?

File: src/main/resources/application.properties [hunk 8]
```diff
@@ -5,7 +5,8 @@ spring.datasource.password=${DB_PASSWORD}
 spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL8Dialect
 spring.jpa.hibernate.ddl-auto=validate
 spring.jpa.show-sql=true
-jwt.expiration=${jwt_expiration}
-jwt.secret=${jwt_secret}
+spring.jpa.properties.hibernate.format_sql=true
+jwt.expiration=3000000000000000000
+jwt.secret=thisismyjvtsectedanditisVERYsercet
```
Comment (karma-o) [inline src/main/resources/application.properties]: let's revert this change

File: src/main/java/com/example/onlinebookstore/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+@SQLRestriction(value = "is_deleted=false")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", referencedColumnName = "id", nullable = false)
+    private User user;
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/model/ShoppingCart.java:39]: Optional: we could also use @MapsId for cart to have the same id as user since we have a 1-1 relation

**Review by karma-o — CHANGES_REQUESTED** [review]
Nice! left a few comments

File: pom.xml (hunk 1 above)
Comment (Eidenn005) [reply pom.xml]: Done

File: src/main/java/com/example/onlinebookstore/model/CartItem.java (hunk 2 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/model/CartItem.java:20]: Done

File: src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java (hunk 3 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java]: Done

File: src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java (hunk 4 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java]: Done

File: src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java (hunk 5 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/service/impl/CartItemServiceImpl.java]: Change method

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 6 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: done

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 7 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: Nope. Delete.

File: src/main/resources/application.properties (hunk 8 above)
Comment (Eidenn005) [reply src/main/resources/application.properties]: done

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+@NoArgsConstructor
+@ToString(exclude = {"shoppingCart", "book"})
+@EqualsAndHashCode(of = {"id"})
```
Comment (karma-o) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java]:
```suggestion
@EqualsAndHashCode(of = {"id", "quantity"})
```

**Review by karma-o — APPROVED** [review]
Well done!

File: src/main/java/com/example/onlinebookstore/model/CartItem.java (hunk 10 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/model/CartItem.java]: Done
